### PR TITLE
[1.10.x] Added a way to create horse armor.

### PR DESCRIPTION
**Pull request by Darkhax**
**The issue being solved**
Horse armor is currently handled through hard references, which makes it impractical for mods to add horse armor. Past attempts usually involve ASM or replacing the horse entity all together, which are not very friendly when it comes to compatibility. 

**How it's being solved**
This PR adds the IHorseArmor interface, which can be implemented by items. This interface gives the item the ability to return a HorseArmorType enum, which can change the texture and protection amount offered. Items with this interface will also be accepted in the horse armor slot. HorseArmorType has been patched to take IHorseArmor into consideration. EntityHorse has also been patched to track the name of the HorseArmorType it has. This allows the related code to retrieve the HorseArmorType by name rather than ordinal. Name lookup is to allow mods to add entries without worrying about load order or being unloaded.

**Use Case**
- Mods will be able to add horse armor, to match the player armor they already have.
- Mods would be able to create custom calvery. Things like Thaumcraft's Crimson Knights riding horses with Crimson armor would be possible.
- Functional horse armor such as horse jet packs or horse wings would be possible. 

**Test Info**
A test mod is included with this PR. The test mod adds a basic piece of horse armor which can be found in the Misc Creative Tab. This armor should give a horse 23 armor points. You can use the following command to spawn a pre-tamed horse at your location. `/summon EntityHorse ~ ~ ~ {Tame:1,SaddleItem:{id:saddle,Count:1}}`

### CODE REVIEW

File: patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
+--- ../src-base/minecraft/net/minecraft/entity/passive/HorseArmorType.java
++++ ../src-work/minecraft/net/minecraft/entity/passive/HorseArmorType.java
+@@ -62,16 +62,19 @@
+ 
+     public static HorseArmorType func_188580_a(@Nullable ItemStack p_188580_0_)
+     {
++    	if (p_188580_0_ != null && p_188580_0_.func_77973_b() instanceof net.minecraftforge.common.IHorseArmor) return ((net.minecraftforge.common.IHorseArmor) p_188580_0_.func_77973_b()).getArmorType(p_188580_0_);
```
Comment (iLexiconn) [inline patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch]: idk what character that is but it's not a space

**Comment by Ellpeck** [comment]
This is a really awesome idea. I could implement this in one of my mods!

File: src/main/java/net/minecraftforge/common/IHorseArmor.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+/**
+ * When implemented by an item, it will be allowed to function as a piece of
+ * horse armor.
+ */
+public interface IHorseArmor 
+{
+
```
Comment (iLexiconn) [inline src/main/java/net/minecraftforge/common/IHorseArmor.java]: also use spaces here

File: src/test/java/net/minecraftforge/test/HorseArmorTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    @SubscribeEvent
+    public void onEntityHit(LivingHurtEvent event) 
+    {
+    	if (event.getEntity() instanceof EntityHorse)
+    		// Armor is limited to 30 by Vanilla
+    		LOG.info("Total Armor: " + event.getEntityLiving().getTotalArmorValue());
+    }
+    public class ItemTestHorseArmor extends Item implements IHorseArmor 
+    {
+    	public ItemTestHorseArmor() {
```
Comment (iLexiconn) [inline src/test/java/net/minecraftforge/test/HorseArmorTest.java]: what's up with the formatting in this class?

**Comment by heldplayer** [comment]
Does this keep compat with vanilla clients when connecting to a Forge server?

**Comment by Darkhax** [comment]
@heldplayer After testing, it seems this PR does break that support. Looks like I will have to find some other way to do textures.

File: src/main/java/net/minecraftforge/common/IHorseArmor.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package net.minecraftforge.common;
```
Comment (kashike) [inline src/main/java/net/minecraftforge/common/IHorseArmor.java:20]: Missing license header.

**Comment by Darkhax** [comment]
Removed the patches to EntityHorse, it seems syncing the worn item will break vanilla support. As there is no way to sync the armor stack, mods will have to use the ItemStack from IHorseArmor#getArmorType to switch textures. I have also added a new constructor to HorseArmorType to allow mods to use a custom domain for their textures.

File: src/main/java/net/minecraftforge/common/IHorseArmor.java (hunk 4 above)
Comment (Darkhax) [inline src/main/java/net/minecraftforge/common/IHorseArmor.java:20]: Added that in. Don't think GitHub will outdate this for me though.

**Comment by LexManos** [comment]
This is not really fasible to do this way as stated it breaks vanilla clients.
Horse with unknown armor == ArrayIndexOutOfBounds.
You can add extra data manager objects to sync to the client and vanilla will happily eat them without any concern.
Similar to how we enhance the Villagers/Zombies.

**Comment by Darkhax** [comment]
That makes sense, didn't account for mods being installed, or load order changes originally. The latest commits use string based lookup, where the string is the name of the constant in `HorseArmorType`. Was thinking about creating a separate registry or lookup specifically for this, however that would likely result in more patching. It is also not necessary to read/write this name from NBT, as it is automatically applied when the armor stack is read from NBT. 

Edit: I have tested this version against load order changes with multiple mods adding horse armor, uninstalling a mod with a certain type of armor, and vanilla client connection. All seem to work as intended.

File: patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch [hunk 5]
```diff
@@ -1,6 +1,40 @@
 --- ../src-base/minecraft/net/minecraft/entity/passive/EntityHorse.java
 +++ ../src-work/minecraft/net/minecraft/entity/passive/EntityHorse.java
-@@ -405,7 +405,7 @@
+@@ -77,6 +77,7 @@
+     private static final DataParameter<Integer> field_184789_bG = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
+     private static final DataParameter<Optional<UUID>> field_184790_bH = EntityDataManager.<Optional<UUID>>func_187226_a(EntityHorse.class, DataSerializers.field_187203_m);
+     private static final DataParameter<Integer> field_184791_bI = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
++    private static final DataParameter<String> ARMOR_STRING = EntityDataManager.<String>func_187226_a(EntityHorse.class, DataSerializers.field_187194_d);
```
Comment (kashike) [inline patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch:7]: Will vanilla clients have an issue with this new entry?

Comment (williewillus) [inline patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch:7]: as long as new DataParameter's are last (aka we do not interfere with the ordering of the original ones), they will be fine and ignored by vanilla clients

Forge adds new DataParameter's to Villagers (or was it zombies?) to support custom variants

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Comment by Darkhax** [comment]
Are there any further issues with this PR?

**Comment by jriwanek** [comment]
Would be useful to me for sure.

**Comment by p455w0rd** [comment]
This would be awesome to finally be able to do

**Comment by Darkhax** [comment]
So, it's been about 3 months since this PR was opened. I know you're busy @LexManos but I would really appreciate knowing your stance on this.

**Comment by jeffryfisher** [comment]
Enabling horse armor has been one of my targets for over two years. If there's work to be done (e.g. testing/debugging), then call on me (git@jeffryfisher.net). If there's a provisional Forge build (for mc 1.10.2) with this feature, then I'd be willing to update a mod ASAP. One of my mods already has a craftable, enchantable horse armor item with texture files; it just doesn't play nice with the slot on the horses.

**Comment by Traverse-Joe** [comment]
Agreed It would Be Amazing

**Comment by Lanse505** [comment]
I +10 this post

**Comment by BunnyCinnamon** [comment]
As I raise my arms /(°^°)/ This shall become a thing!

**Comment by HenryLoenwind** [comment]
Good thing to have. And while you're at it, you can continue with the Shield and the Elytra. ;)

File: patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+     }
+ 
+     public static boolean func_188577_b(@Nullable Item p_188577_0_)
+     {
++        if (p_188577_0_ instanceof net.minecraftforge.common.IHorseArmor) return true;
+         return func_188576_a(p_188577_0_) != NONE;
+     }
++    
++    // Forge    
++    private HorseArmorType(int armorStengthIn, String textureName)
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:24]: There is nothing in here deprecating the old methods, or killing off the usages of the old bad methods.
The constructor also has no documentation. 
Also by preliminary glance at how this class is used, it looks like you need some mechanic of specifying the domain for the texture in EntityHorse.setHorseTexturePaths

File: src/main/java/net/minecraftforge/common/IHorseArmor.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+import net.minecraft.entity.passive.HorseArmorType;
+import net.minecraft.item.ItemStack;
+
+/**
+ * When implemented by an item, it will be allowed to function as a piece of
+ * horse armor.
+ */
+public interface IHorseArmor
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/IHorseArmor.java:29]: not a fan of single use interfaces like this... it definitely shouldn't just be slapped in common. It ONLY ever gunna be applicable to items, why not just make it a thing on item?

File: src/main/java/net/minecraftforge/common/IHorseArmor.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+public interface IHorseArmor 
+{
+    /**
+     * Provides the HorseArmorType to use for this armor. Used to change the
+     * protection value provided.
+     * 
+     * @param stack The ItemStack instance of the armor.
+     * @return HorseArmorType The HorseArmorType to use for this armor.
+     */
+    HorseArmorType getArmorType(ItemStack stack);
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/IHorseArmor.java:38]: Needs a @Nullable as this is called in places where you only have the Item and not the stack, unless you go through and fix that.

File: src/main/java/net/minecraftforge/common/util/EnumHelper.java [hunk 9]
```diff
@@ -123,6 +125,10 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
     {
         return addEnum(EnumRarity.class, name, color, displayName);
     }
+    public static HorseArmorType addArmorType(String name, int protection, String texture)
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/util/EnumHelper.java:128]: no docs no nothing esp on a constructor added by Forge.

**Comment by LexManos** [comment]
Looks like you've missed A LOT of the places where armor type is used to determine things.
You're lacking documentation, and the single purpose interfaces are not an acceptible design for Forge.
Its fine in other things but we've gone the route of just putting them as methods instead of constantly having to do InstanceOf

**Comment by mezz** [comment]
Please update this to address Lex's comments and to target 1.11, new features are no longer being merged into 1.10.

File: src/main/java/net/minecraftforge/common/IHorseArmor.java (hunk 7 above)
Comment (liach) [reply src/main/java/net/minecraftforge/common/IHorseArmor.java:29]: This should be changed into capabilities possibly.

**Comment by mezz** [comment]
Any update @darkhax ?

**Comment by mezz** [comment]
I am closing this since you have not responded in a long time.
If you decide to update this please open a new PR and link back to this one.

**Pull request closed without merging**